src/charm.py

```diff
@@ -24,37 +27,37 @@ def __init__(self, *args: typing.Any):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
-        self.irc = IRCService()
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.install, self._on_install)
-        self.framework.observe(self.on.start, self._on_start)
+        self._irc = IRCService()
+        self._database = DatabaseService(self, DATABASE_RELATION_NAME)
+        self.framework.observe(
+            self._database.database.on.database_created, self.reconcile
+        )
+        self.framework.observe(
+            self.on[DATABASE_RELATION_NAME].relation_broken,
+            self.reconcile
+        )
+        self.framework.observe(self.on.config_changed, self.reconcile)
+        self.framework.observe(self.on.install, self.reconcile)
+        self.framework.observe(self.on.start, self.reconcile)
         self.framework.observe(self.on.stop, self._on_stop)
-        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
 
-    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
-        """Handle changed configuration."""
-        self.unit.status = ops.ActiveStatus()
+    def reconcile(self) -> None:
+        """Reconcile the charm.
+        This is a more simple approach to reconciliation,
+        adapted from Charming Complexity sans state and observers.
 
-    def _on_install(self, _: ops.InstallEvent) -> None:
-        """Handle install."""
-        self.unit.status = ops.MaintenanceStatus("Preparing irc bridge")
-        self.irc.prepare()
-        self.unit.status = ops.ActiveStatus()
+        Being a simple charm, we don't need to do much here.
 
-    def _on_start(self, _: ops.StartEvent) -> None:
-        """Handle start."""
-        self.irc.start()
-        self.unit.status = ops.ActiveStatus()
+        Ensure we have a database relation,
+        ensure we have a relation to matrix,
+        populate database connection string and matrix homeserver URL
+        in the config template and (re)start the service.
+        """
+        self._irc.reconcile()
```

Comment:
I guess the checks in the comments (database/relation) will still be added, right?

Reply:
I added the DB checks in the src/irc.py and the charm here just calls the big thing, wdyt? The comments could use adjusting, I agree, I place them as guidelines while this is still in draft.